source/code/providers/support/mycnf-scanner.cpp

```diff
@@ -0,0 +1,244 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * ---------------------------------- END OF LICENSE -----------------------------
+*/
+/**
+    \file        mycnf-scanner.cpp
+
+    \brief       Scans mysql configuration to extract authentication information.
+
+    \date        2014-11-05 16:22:00
+
+*/
+/*----------------------------------------------------------------------------*/
+
+#include "mycnf-scanner.h"
+
+namespace MySQL
+{
+    static inline bool GetPathFromDirective(const std::wstring& line, SCXCoreLib::SCXFilePath& path)
+    {
+        std::vector<std::wstring> elements;
+        SCXCoreLib::StrTokenize(line, elements, L" ");
+        if ( elements.size() != 2 )
+        {
+            return false;
+        }
+
+        path.Set(elements[1]);
+        return true;
+    }
+
+    bool MySQL_ConfigurationScanner::LoadCnfFile(
+        const SCXCoreLib::SCXFilePath& file, 
+        MySQL_Authentication& auth,
+        std::set< std::wstring >& visited, 
+        std::map<unsigned int, std::wstring>& ports)
+    {
+        SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" loading: ", file.Get()));
+        std::pair<std::set< std::wstring >::iterator, bool> ins = visited.insert(file.Get());
+        if (!ins.second)
+        {
+            std::wstringstream ss;
+            ss << L"Already visited " << file.Get() << L". skipping";
+            SCX_LOGWARNING(m_log, ss.str());
+            return true;
+        }
+
+        std::vector< std::wstring > lines;
+        try
+        {
+            SCXCoreLib::SCXStream::NLFs nlfs;
+
+            m_extDepends->ReadAllLinesAsUTF8(file, lines, nlfs);
+        }
+        catch (SCXCoreLib::SCXException & e)
+        {
+            std::wstringstream ss;
+            ss << L"Caught " << e.What() << L" when trying to read " << file.Get() ;
+            SCX_LOGERROR(m_log, ss.str());
+            return false;
+        }
+
+        unsigned int portNum = 0;
+        std::wstring bindAddress;
+        for(std::vector< std::wstring >::const_iterator iter = lines.begin(); iter != lines.end(); ++iter)
+        {
+            std::wstring line( SCXCoreLib::StrTrim(*iter) );
+            SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" Line: ", line));
+            if (SCXCoreLib::StrIsPrefix(line, L"!includedir"))
+            {
+                SCXCoreLib::SCXFilePath directory;
+                if (GetPathFromDirective(line, directory))
+                {
+                    LoadCnfDirectory(directory, auth, visited, ports);
+                }
+                else
+                {
+                    SCX_LOGWARNING(m_log, SCXCoreLib::StrAppend(L" Ignoring Error in !includedir directive.", line));
```

Comment:
I'm not sure what you're saying is true.  What if I have "!includedir foo bar".  Would "foo bar" be interpreted as a directory (remember, while "bad form", spaces are allowed in directory names)? If so, that would make this a legal statement.

In general, on lines that we parse, if we find something invalid, we should report it right there and then and probably stop reading. This makes it easier to find parsing code down the road for specific file sections and change parsing as necessary.

In this particular case, the parser is still pretty "dumb"; there's a lot of lines we just don't care about and ignore. But for what we do care about, if we find something we don't like, we should just log and abort.

---
In reply to: [66897123](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66897123) [](ancestors = 66897123)